File: app/src/main/java/com/sies/movierecomendations/MoviesRecycler/MovieListAdapter.java
package com.sies.movierecomendations.MoviesRecycler;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.sies.movierecomendations.MoviesApi.MoviesList;
import com.sies.movierecomendations.R;

import java.util.HashMap;
import java.util.Map;

public class MovieListAdapter extends RecyclerView.Adapter<MovieListAdapter.ViewHolder> {

    FirebaseDatabase database = FirebaseDatabase.getInstance();
    DatabaseReference mDatabase = database.getReference();
    FirebaseUser person = FirebaseAuth.getInstance().getCurrentUser();

    MoviesList res;
    int count = 0, fav = 0;
    private Context context;
    public MovieListAdapter (MoviesList res) {
        this.res = res;
    }

    @NonNull
    @Override
    public MovieListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.movie_list, parent, false);
        context = parent.getContext();
        return new ViewHolder(view);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull MovieListAdapter.ViewHolder holder, int position) {
        Glide.with(holder.itemView.getContext()).load("https://image.tmdb.org/t/p/w500"+ res.getResults().get(position).getBackdrop_path()).into(holder.poster);
        holder.title.setText(res.getResults().get(position).getTitle());
        holder.desc.setText(res.getResults().get(position).getOverview());
        holder.frame.setOnClickListener(v -> {
            if(count%2==0) holder.desc.setVisibility(View.VISIBLE);
            else holder.desc.setVisibility(View.GONE);
            count++;
        });
        holder.date.setText("Release Date: " + res.getResults().get(position).getRelease_date());
        holder.star.setOnClickListener(v -> {
            int pos;
            pos = position;
            Toast.makeText(context, Integer.toString(pos), Toast.LENGTH_SHORT).show();
            if(fav%2==0) {
                holder.star.setImageResource(R.drawable.ic_baseline_star_rate_24);
                Map<String, Object> childUpdates = new HashMap<>();
                childUpdates.put("id", res.getResults().get(position).getTitle());
                mDatabase.child("Users").child(person.getUid()).child("favourites").updateChildren(childUpdates);
            }
            else holder.star.setImageResource(R.drawable.ic_baseline_star_outline_24);
            fav++;
        });
//        Map<String, Object> childUpdates = new HashMap<>();
//        childUpdates.put("id", res.getResults().get(position).getId());
//        mDatabase.child("Users").child(person.getUid()).child("favourites").updateChildren(childUpdates);
    }

    @Override
    public int getItemCount() {
        return res.getResults().size();
    }
    public class ViewHolder extends RecyclerView.ViewHolder {

        private ImageView poster;
        private TextView title;
        private TextView desc;
        private TextView date;
        private ConstraintLayout frame;
        private ImageView star;
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            poster = itemView.findViewById(R.id.image);
            title = itemView.findViewById(R.id.title);
            desc = itemView.findViewById(R.id.description);
            date = itemView.findViewById(R.id.releaseDate);
            frame = itemView.findViewById(R.id.frame);
            star = itemView.findViewById(R.id.star);
        }

    }
}

File: app/src/main/java/com/sies/movierecomendations/Genre/ui/MovieListAdapter.kt
package com.sies.movierecomendations.Genre.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.sies.movierecomendations.databinding.GenreMoviesCardBinding
import com.sies.movierecomendations.network.Results

class MovieListAdapter(private val clickListener: MovieListClickListener): ListAdapter<Results, MovieListAdapter.ViewHolder>(MovieListDiffCallback()) {
    class ViewHolder(private var binding: GenreMoviesCardBinding): RecyclerView.ViewHolder(binding.root) {
        fun bind(res: Results, clickListener: MovieListClickListener) {
            binding.movies = res
//            binding.clickListener = clickListener
            binding.frame.setOnClickListener {
                clickListener.onClick(res)
            }
            binding.executePendingBindings()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(GenreMoviesCardBinding.inflate(LayoutInflater.from(parent.context)))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val results = getItem(position)
        holder.bind(results, clickListener)
    }

}

class MovieListClickListener(val clickListener: (res: Results) -> Unit) {
    fun onClick(res: Results) = clickListener(res)
}

class MovieListDiffCallback: DiffUtil.ItemCallback<Results>() {
    override fun areItemsTheSame(oldItem: Results, newItem: Results): Boolean {
        return oldItem.id == newItem.id
    }
    override fun areContentsTheSame(oldItem: Results, newItem: Results): Boolean {
        return oldItem.equals(newItem)
    }

}
